src/masternodes/rpc_vault.cpp

```diff
@@ -1146,6 +1146,105 @@ UniValue listauctionhistory(const JSONRPCRequest& request) {
     return ret;
 }
 
+UniValue estimatecollateral(const JSONRPCRequest& request) {
+    auto pwallet = GetWallet(request);
+
+    RPCHelpMan{"estimatecollateral",
+               "Returns amount of collateral tokens needed to take an amount of loan tokens for a target collateral ratio.\n",
+                {
+                    {"loanAmounts", RPCArg::Type::STR, RPCArg::Optional::NO,
+                        "Amount as json string, or array. Example: '[ \"amount@token\" ]'"
+                    },
+                    {"targetRatio", RPCArg::Type::NUM, RPCArg::Optional::NO, "Target collateral ratio."},
+                    {"tokens", RPCArg::Type::OBJ, RPCArg::Optional::OMITTED, "Object with collateral token as key and their percent split as value. (defaults to { DFI: 1 }",
+                        {
+                            {"split", RPCArg::Type::NUM, RPCArg::Optional::NO, "The percent split"},
+                        },
+                    },
+                },
+                RPCResult{
+                    "\"json\"                  (Array) Array of <amount@token> strings\n"
+                },
+               RPCExamples{
+                       HelpExampleCli("estimatecollateral", R"(23.55311144@MSFT 150 '{"DFI": 0.8, "BTC":0.2}')") +
+                       HelpExampleRpc("estimatecollateral", R"("23.55311144@MSFT" 150 {"DFI": 0.8, "BTC":0.2})")
+               },
+    }.Check(request);
+
+    RPCTypeCheck(request.params, {UniValueType(), UniValue::VNUM, UniValue::VOBJ}, false);
+
+    const CBalances loanAmounts = DecodeAmounts(pwallet->chain(), request.params[0], "");
+    auto ratio = request.params[1].get_int();
+
+    std::map<std::string, UniValue> collateralSplits;
+    if (request.params.size() > 2) {
+        request.params[2].getObjMap(collateralSplits);
+    } else {
+        collateralSplits["DFI"] = 1;
+    }
+
+    LOCK(cs_main);
+
+    CAmount totalLoanValue{0};
+    for (const auto& [tokenId, tokenAmount] : loanAmounts.balances) {
+        auto loanToken = pcustomcsview->GetLoanTokenByID(tokenId);
+        if (!loanToken) {
+            throw JSONRPCError(RPC_DATABASE_ERROR, strprintf("(%s) is not a loan token!", tokenId));
```

Comment:
loanToken->symbol is better fit? At least tokenId.v is int not string.

Reply:
Cannot get `loanToken->symbol` as `!loanToken` but fixed it using `tokenId.v` as int